x/gamm/twap/api_test.go

```diff
@@ -0,0 +1,204 @@
+	initStartRecord.PoolId, initStartRecord.Height = poolId, s.Ctx.BlockHeight()
+	initStartRecord.Asset0Denom, initStartRecord.Asset1Denom = denomA, denomB
+
+	zeroAccumTenPoint1Record := recordWithUpdatedSpotPrice(initStartRecord, sdk.NewDec(10), sdk.NewDecWithPrec(1, 1))
+
+	blankRecord := types.TwapRecord{}
+	defaultTime := s.Ctx.BlockTime()
+
+	tPlusOneSec := defaultTime.Add(time.Second)
+
+	tests := map[string]struct {
+		// if start record is blank, don't do any sets
+		startRecord types.TwapRecord
+		// We set it to have the updated time
+		expRecord  types.TwapRecord
+		time       time.Time
+		poolId     uint64
+		quoteDenom string
+		baseDenom  string
+		expError   bool
+	}{
+		"no record (wrong pool ID)": {blankRecord, blankRecord, defaultTime, 4, denomA, denomB, true},
+		"default record":            {blankRecord, initStartRecord, defaultTime, 1, denomA, denomB, false},
+		"one second later record":   {blankRecord, recordWithUpdatedAccum(initStartRecord, OneSec, OneSec), tPlusOneSec, 1, denomA, denomB, false},
+		"idempotent overwrite":      {initStartRecord, initStartRecord, defaultTime, 1, denomA, denomB, false},
+		"idempotent overwrite2":     {initStartRecord, recordWithUpdatedAccum(initStartRecord, OneSec, OneSec), tPlusOneSec, 1, denomA, denomB, false},
+		"diff spot price": {zeroAccumTenPoint1Record,
+			recordWithUpdatedAccum(zeroAccumTenPoint1Record, OneSec.MulInt64(10), OneSec.QuoInt64(10)),
+			tPlusOneSec, 1, denomA, denomB, false},
+		// TODO: Overflow
+	}
+	for name, tc := range tests {
+		s.Run(name, func() {
+			// setup time
+			s.Ctx = s.Ctx.WithBlockTime(tc.time)
+			tc.expRecord.Time = tc.time
+
+			// setup record
+			initSetRecord := tc.startRecord
+			if (tc.startRecord == types.TwapRecord{}) {
+				initSetRecord = initStartRecord
+			}
+			s.twapkeeper.StoreNewRecord(s.Ctx, initSetRecord)
+
+			actualRecord, err := s.twapkeeper.GetBeginBlockAccumulatorRecord(s.Ctx, tc.poolId, tc.baseDenom, tc.quoteDenom)
+
+			if tc.expError {
+				s.Require().Error(err)
+				return
+			}
+			s.Require().NoError(err)
+			s.Require().Equal(tc.expRecord, actualRecord)
+		})
+	}
+}
+
+func (s *TestSuite) TestGetArithmeticTwap() {
+	type getTwapInput struct {
+		poolId          uint64
+		quoteAssetDenom string
+		baseAssetDenom  string
+		startTime       time.Time
+		endTime         time.Time
+	}
+
+	makeSimpleTwapInput := func(startTime time.Time, endTime time.Time, isQuoteTokenA bool) getTwapInput {
+		quoteAssetDenom, baseAssetDenom := denom0, denom1
+		if isQuoteTokenA {
+			baseAssetDenom, quoteAssetDenom = quoteAssetDenom, baseAssetDenom
+		}
+		return getTwapInput{1, quoteAssetDenom, baseAssetDenom, startTime, endTime}
+	}
+
+	quoteAssetA := true
+	quoteAssetB := false
+	tPlusOne := baseTime.Add(time.Second)
+	// base record is a record with t=baseTime, sp0=10, sp1=.1, accumulators set to 0
+	baseRecord := newTwapRecordWithDefaults(baseTime, sdk.NewDec(10), sdk.ZeroDec(), sdk.ZeroDec())
```

Comment:
I've seen `baseRecord` and `tPlusOne` being used in other tests. Can we reuse these in the test-global scope?

Reply:
Done!